sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuShuffleExchangeExecBase.scala

```diff
@@ -417,9 +419,25 @@ object GpuShuffleExchangeExecBase {
               // Get a non-empty batch or the last batch. So still need to
               // check if it is empty for the later case.
               if (batch.numRows > 0) {
+
+                // Check if this is the only batch. We do it before partitioning
+                // because partitioning may release the semaphore, making it time
+                // consuming to do hasNext check.
+                if (isFirstCheck) {
+                  isFirstCheck = false
+
+                  if (!iter.hasNext) {
```

Comment:
This makes me concerned from a memory pressure standpoint. We don't know what is upstream of this node so asking if there is more could require us to fully materialize the next batch.  We do this in the window operator when we know that sort is the thing feeding into it so asking it is free.

Could we add a comment here that this is possible to happen, but the performance win outweighs the cost.